lib/Doctrine/ODM/MongoDB/Mapping/ClassMetadata.php

```diff
@@ -1220,9 +1220,15 @@ public function hasIndexes(): bool
      */
     public function addSearchIndex(array $definition, ?string $name = null): void
     {
+        $name ??= self::DEFAULT_SEARCH_INDEX_NAME;
+
+        if (empty($definition['mappings']['dynamic']) && empty($definition['mappings']['fields'])) {
```

Comment:
Using `array_key_exists()` here would most directly address the original bug. Both [XmlDriver](https://github.com/doctrine/mongodb-odm/blob/2.9.0/lib/Doctrine/ODM/MongoDB/Mapping/Driver/XmlDriver.php#L583) an [Attribute](https://github.com/doctrine/mongodb-odm/blob/2.9.0/lib/Doctrine/ODM/MongoDB/Mapping/Driver/AttributeDriver.php#L358) driver should only set these keys if the mappings actually specified a value, so I don't think we need to prevent users from specifying `dynamic: false` and empty fields.

Reply:
I want to check if `$definition['mappings']['dynamic']` is `false` also. 
The condition would be:

```suggestion
        if (! array_key_exists('mappings', $definition) || ((! array_key_exists('dynamics', $definition['mappings']) || ! $definition['mappings']) && (! array_key_exists('fields', $definition['mappings']) || ! is_array($definition['mappings']) || count($definition['mappings']) === 0)) {
```

It don't think this makes the code more explicit, secure or performant.

Reply:
My understanding was that the original bug only cared about both options being unspecified, so two `array_key_exists()` checks would be the minimum change to implement. ODM wasn't prohibiting `dynamic: true` with a non-empty `fields` array (I assume the server reports an error there), so I assumed the most minimal validation would be OK.

If you instead want to prohibit `dynamic: false` and an empty `fields` array, the original condition was fine (and certainly more readable). Since we know the server doesn't report an error for that, I suppose having ODM validate this will protect users from creating a useless index.

Reply:
I totally messed with github UI patches. I'll revert this change.